src/beamlime/communication/pipes.py

```diff
@@ -0,0 +1,238 @@
+    """
+
+    def __init__(
+        self,
+        *_initial_data: BufferData,
+        max_size: int = MAX_BUFFER_SIZE,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+    ) -> None:
+        super().__init__(*_initial_data, max_size=max_size)
+        self.chunk_size: int = chunk_size
+
+    @property
+    def chunk_size(self) -> int:
+        return self._chunk_size
+
+    @chunk_size.setter
+    def chunk_size(self, _chunk_size: int) -> None:
+        if _chunk_size > MAX_CHUNK_SIZE:
+            raise ValueError("Maximum chunk size is limited to ", MAX_CHUNK_SIZE)
+        self._chunk_size = _chunk_size
+
+    @contextmanager
+    def _open_buffer(
+        self, buffer_type: BufferType[BufferData]
+    ) -> Iterator[Buffer[BufferData]]:
+        """
+        Create a buffer of the ``buffer_type``.
+        """
+        try:
+            from copy import deepcopy
+
+            chunk, self._data = (
+                self._data[: self.chunk_size],
+                self._data[self.chunk_size :],
+            )
+            _buffer = buffer_type(*deepcopy(chunk), max_size=self.chunk_size)
+            yield _buffer
+        except Exception as err:
+            self._data = chunk + self._data
+            raise err
+        else:
+            consumed = max(0, len(chunk) - len(_buffer))
+            self._data = chunk[consumed:] + self._data
+        finally:
+            ...
+
+    @contextmanager
+    def open_readable(
+        self, timeout: int = 0, retry_interval: float = 0.1
+    ) -> Iterator[BufferBase[BufferData]]:
+        """
+        Yield a buffer containing a copy of the first chunk of data.
+        """
+        from ..core.schedulers import retry
+
+        max_trials = int(timeout / retry_interval) + 1
+
+        @retry(Empty, max_trials=max_trials, interval=retry_interval)
+        def wait_for_data() -> None:
+            if not self._data:
+                raise Empty("There is no data left in the pipe.")
+
+        wait_for_data()
+        try:
+            with self._open_buffer(SyncBuffer[BufferData]) as _buffer:
+                yield _buffer
+        finally:
+            ...
+
+    @asynccontextmanager
+    async def open_async_readable(
+        self, timeout: int = 0, retry_interval: float = 0.1
+    ) -> AsyncIterator[BufferBase[BufferData]]:
+        """
+        Yield an async buffer containing a copy of the first chunk of data.
+        """
+
+        from ..core.schedulers import async_retry
+
+        max_trials = int(timeout / retry_interval) + 1
+
+        @async_retry(Empty, max_trials=max_trials, interval=retry_interval)
+        async def wait_for_data() -> None:
+            if not self._data:
+                raise Empty("There is no data left in the pipe.")
+
+        await wait_for_data()
+
+        try:
+            with self._open_buffer(AsyncBuffer[BufferData]) as _buffer:
+                yield _buffer
+        finally:
+            ...
+
+    def write(self, data: BufferData) -> None:
```

Comment:
Is this meant to be written to from different threads? I can't see any locking mechanism (or other) to make this safe?

Reply:
That is why it has `open_readable_*`  interface.
So it does not guarantee the order of data, but it will not lose any data that are not processed by the consumer.
I thought it'll be enough since `kafka` doesn't guarantee the order of data either.

Reply:
I am not entirely convinced it is so easy to support concurrent writes and reads safely? Or I don't understand the mechanism.

Reply:
If I explain what it does more in detail...

When a consumer reads a buffer from the context manager,
`Pipe` splits a chunk from the original list and give the chunk to the consumer.
Once the context is exited without any error, it discard the chunk and the original list stays the same.
If there is any error raised within the context, it returns the chunk into the original list.

But writing doesn't matter since it will not `write` until the data is ready, so it is not a context manager, it's just appending interface.


Do you think it's not enough to say it's `safe` or is it too handling error too much...?